---
title: "Differential expression analysis for platelet gene in different cohort"
author: "Romane Cathelin"
output: html_document
---


```{r libraries}
packages.list <- c("dplyr", "tibble", "ggplot2", "DESeq2", "kableExtra", "viridis")
for (package in packages.list) {
  if (!require(package, character.only = TRUE)) {
    install.packages(package)
  }
}
sapply(packages.list, library, character.only = TRUE)
source("/gpfs/home/cather01/platelet/matt_deseq_function.R")
source("/gpfs/home/cather01/platelet/PLATELET/plot_functions.R")

```

```{r paths}
input.path <- "/gpfs/data/abl/home/cather01/rugglesLab/PACE/DATA/"
output.path <- "/gpfs/data/abl/home/cather01/rugglesLab/PACE/"
```

```{r young_vs_old}
#### Young vs Old ###
data = read.csv(paste0(input.path, "countmatrix_plt_PACE.csv"), row.names = 1)
metadata = read.csv(paste0(input.path, "pace_and_thr_metadata.csv"), row.names = 1)
metadata <- metadata[rownames(metadata) %in% colnames(data), ]

# check the distribution 
p <- ggplot(metadata, aes(x=age)) + 
  geom_histogram(fill="#69b3a2", color="#e9ecef", alpha=0.9, bins = 20) +
  theme_bw() +
  scale_x_continuous(breaks = seq(0, 100, 5)) +
  labs(title = "Distribution of Ages",
       x = "Age",
       y = "Frequency") +
  theme_minimal()

# Delete outliers 
metadata <- as.data.frame(metadata[rownames(metadata) != "PACE011", ])
data <- data[, colnames(data) != "PACE011"]

#Transform age in young and old
metadata <- metadata %>%
  mutate(age_group = ifelse(age < 73, "young", "old"))

#Transform column into factor
met <- as.data.frame(metadata$age_group)
rownames(met)<-rownames(metadata)
colnames(met) <- c("age")
met$age <- as.factor(met$age)


young_data <- met[met$age == "young", ]
old_data <- met[met$age == "old", ]
length(young_data)
length(old_data)


#Verification
all(colnames(data) %in% rownames(met))
all(colnames(data) == rownames(met))

#Construct DESEQ2 object
dds <- DESeqDataSetFromMatrix(countData = data, 
                               colData = met,
                               design = ~ age)
dds


#pre filtering --> removing rows with low gene counts
# keeping rows that have at least 10 reads total
dds2 <- rna_preprocessing(dds, paste0(output.path,"plt_deseq_preprocess_age"))


#Set the factor level (3 level = 3 categories d'ages)
dds2$age <- relevel(dds$age, ref = "young")


#Run DESeq
dds2 <- DESeq(dds2)
res2 <- results(dds2)
summary(res2)
# saveRDS(dds2, "/gpfs/data/abl/home/cather01/rugglesLab/plt_vs_wb/dds_plt.rds")

#Check how many gene are significant
sum(res2$padj < 0.2, na.rm=TRUE)
length(which(res2$pvalue < 0.05))

resultsNames(dds2)
res_tableOE <- lfcShrink(dds2, coef="age_old_vs_young")


volcanoplot(dds2,  "Young patient (<68) vs old", 0.5, 0.2)
plot_heatmap_zscores_categories(dds2, 0.5, 0.1)
plot_heatmap_log2FC_categories(dds2, 0.5, 0.1)


# Error in results(dds) : couldn't find results. you should first run DESeq()
```




```{r male_vs_female}
#### Male vs Female###
data = read.csv(paste0(input.path, "countmatrix_plt_PACE.csv"), row.names = 1)
metadata = read.csv(paste0(input.path, "pace_and_thr_metadata.csv"), row.names = 1)
metadata <- metadata[rownames(metadata) %in% colnames(data), ]

names(metadata)
metadata$patient <- rownames(metadata)

# Delete outliers 
metadata <- as.data.frame(metadata[rownames(metadata) != "PACE011", ])
data <- data[, colnames(data) != "PACE011"]


#Transform column into factor
met <- as.data.frame(metadata$sex)
rownames(met)<-rownames(metadata)
colnames(met) <- c("sex")
met$sex <- as.factor(met$sex)


female_data <- met[met$sex == "Female", ]
male_data <- met[met$sex == "Male", ]
length(female_data)
length(male_data)


#Verification
all(colnames(data) %in% rownames(met))
all(colnames(data) == rownames(met))

#Construct DESEQ2 object
dds <- DESeqDataSetFromMatrix(countData = data, 
                               colData = met,
                               design = ~ sex)
dds

# Filtering 
dds <- rna_preprocessing(dds, paste0(output.path,"plt_deseq_preprocess_age"))

# Run DESeq
dds <- DESeq(dds)
res <- results(dds)
summary(res)

sum(res$padj < 0.2, na.rm=TRUE)


volcanoplot(dds,  "Male vs female patient", 0.5, 0.2)
plot_heatmap_zscores_categories(dds, 2, 0.05)
plot_heatmap_log2FC_categories(dds, 2, 0.05)


## Filter out outliers
filtered_res.df <- res.df[!(abs(res.df$log2FoldChange) > 10 | res.df$pvalue > 1e-10), ]


```



```{r else}

out.path <- "/gpfs/data/abl/home/cather01/rugglesLab/"
tabular.stat.table <- metadata %>%
  dplyr::select("age", "sex", "ethnicity", "smoking", "bmi", "diabetes","CAD","hypertension", "hyperlipidemia") %>%
  kbl("html") %>%
  kable_styling(bootstrap_options = "striped", full_width = F) %>%
  kable_minimal() %>%
  save_kable(paste0(out.path, "stat_table.html"))

# Filtering 
dds2 <- rna_preprocessing(dds, "/gpfs/data/abl/home/cather01/rugglesLab/PACE/plt_deseq_preprocess")
dds2
# Plot histogram with ggplot2
#Run DESeq
dds2 <- DESeq(dds2)
res2 <- results(dds2)
summary(res2)

max(res2$lfcSE)
res2 <- res2[res2$log2FoldChange < 9.947548,]
# saveRDS(dds2, "/gpfs/data/abl/home/cather01/rugglesLab/plt_vs_wb/dds_plt.rds")

#Check how many gene are significant
sum(res2$padj < 0.1, na.rm=TRUE)
length(which(res2$pvalue < 0.01))


## MA plot 
DESeq2::plotMA(res2)


resultsNames(dds2)
res_tableOE <- lfcShrink(dds2, coef="sex_Male_vs_Female")


DESeq2::plotMA(res2)
DESeq2::plotMA(res_tableOE)

res_tableOE_tb <- res_tableOE %>%
  data.frame() %>%
  rownames_to_column(var="gene") %>% 
  as_tibble()

sigOE <- res_tableOE_tb %>%
        filter(padj < 0.2 & abs(log2FoldChange) > 0.45)

list.gene <- get_fc_list(res2)
df.enri <- rna_enrichment(list.gene,  "/gpfs/data/abl/home/cather01/rugglesLab/PACE/enrich")
df.gsea <- gsea_analysis(list.gene,  "/gpfs/data/abl/home/cather01/rugglesLab/PACE/gsea_analysis")
```

