---
title: "Differential expression analysis for platelet gene in different cohort"
author: "Romane Cathelin"
output: html_document
---


```{r libraries}
packages.list <- c("dplyr", "tibble", "ggplot2", "DESeq2", "kableExtra", "viridis")
for (package in packages.list) {
  if (!require(package, character.only = TRUE)) {
    install.packages(package)
  }
}
sapply(packages.list, library, character.only = TRUE)
source("/gpfs/home/cather01/platelet/matt_deseq_function.R")

```

```{r paths}
input.path <- "/gpfs/data/abl/home/cather01/rugglesLab/PACE/DATA/"
output.path <- "/gpfs/data/abl/home/cather01/rugglesLab/PACE/"
```

```{r young_vs_old}
#### Young vs Old ###
data = read.csv(paste0(input.path, "countmatrix_plt_PACE.csv"), row.names = 1)
metadata = read.csv(paste0(input.path, "pace_and_thr_metadata.csv"), row.names = 1)
metadata <- metadata[rownames(metadata) %in% colnames(data), ]

# check the distribution 
p <- ggplot(metadata, aes(x=age)) + 
  geom_histogram( fill="#69b3a2", color="#e9ecef", alpha=0.9)+theme_bw()+ scale_x_continuous(breaks = seq(0, 100, 5))

# Delete outliers 
metadata <- as.data.frame(metadata[rownames(metadata) != "PACE011", ])
data <- data[, colnames(data) != "PACE011"]

#Transform age in young and old
metadata <- metadata %>%
  mutate(age_group = ifelse(age < 68, "young", "old"))

#Transform column into factor
met <- as.data.frame(metadata$age_group)
rownames(met)<-rownames(metadata)
colnames(met) <- c("age")
met$age <- as.factor(met$age)


young_data <- met[met$age == "young", ]
old_data <- met[met$age == "old", ]
length(young_data)
length(old_data)


#Verification
all(colnames(data) %in% rownames(met))
all(colnames(data) == rownames(met))

#Construct DESEQ2 object
dds <- DESeqDataSetFromMatrix(countData = data, 
                               colData = met,
                               design = ~ age)
dds


#pre filtering --> removing rows with low gene counts
# keeping rows that have at least 10 reads total
dds2 <- rna_preprocessing(dds, "/gpfs/data/abl/home/cather01/rugglesLab/PACE/plt_deseq_preprocess")


#Set the factor level (3 level = 3 categories d'ages)
dds2$age <- relevel(dds$age, ref = "young")


#Run DESeq
dds2 <- DESeq(dds2)
res2 <- results(dds2)
summary(res2)
# saveRDS(dds2, "/gpfs/data/abl/home/cather01/rugglesLab/plt_vs_wb/dds_plt.rds")

#Check how many gene are significant
sum(res2$padj < 0.2, na.rm=TRUE)
length(which(res2$pvalue < 0.01))


## MA plot 
plotMA(res2)

# Volcano plot
volcano_plot <- function(res2, threshold = 0.01, log2FC_threshold = 1) {
  significant <- res2$pvalue < threshold & abs(res2$log2FoldChange) > log2FC_threshold
  plot(res2$log2FoldChange, -log10(res2$pvalue), pch=20, cex=0.5, col=ifelse(significant, "red", "black"),
       main="Volcano Plot", xlab="log2 Fold Change", ylab="-log10(p-value)")
  abline(h = -log10(threshold), col="blue", lty=2)
  abline(v = c(-log2FC_threshold, log2FC_threshold), col="blue", lty=2)
}

# Create the volcano plot
volcano_plot(res2)


```


```{r}
#### Young vs Old ###
data = read.csv("/gpfs/data/abl/home/cather01/rugglesLab/plt_vs_wb/countmaxtrix_plt.csv", row.names = 1)
metadata = read.csv("/gpfs/data/abl/home/cather01/rugglesLab/PACE/pace_and_thr_metadata.csv", row.names = 1)
metadata <- metadata[rownames(metadata) %in% colnames(data), ]

p <- ggplot(metadata, aes(x=age)) + 
  geom_histogram( fill="#69b3a2", color="#e9ecef", alpha=0.9)+theme_bw()+ scale_x_continuous(breaks = seq(0, 100, 5))


# Delete outliers 
metadata <- as.data.frame(metadata[rownames(metadata) != "PACE011", ])
data <- data[, colnames(data) != "PACE011"]

#Transform age in young and old
metadata <- metadata %>%
  mutate(age_group = ifelse(age < 64, "young", "old"))


#Transform column into factor
met <- as.data.frame(metadata$age_group)
rownames(met)<-rownames(metadata)
colnames(met) <- c("age")
met$age <- as.factor(met$age)


unique(met$age)
young_data <- met[met$age == "young", ]
old_data <- met[met$age == "old", ]
length(young_data)
length(old_data)


#Verification
all(colnames(data) %in% rownames(met))
all(colnames(data) == rownames(met))

#Construct DESEQ2 object
dds <- DESeqDataSetFromMatrix(countData = data, 
                               colData = met,
                               design = ~ age)
dds


# Filtering 
dds2 <- rna_preprocessing(dds, "/gpfs/data/abl/home/cather01/rugglesLab/PACE/plt_deseq_preprocess_age")
dds2
# Plot histogram with ggplot2
#Run DESeq
dds2 <- DESeq(dds2)
res2 <- results(dds2)
summary(res2)
# saveRDS(dds2, "/gpfs/data/abl/home/cather01/rugglesLab/plt_vs_wb/dds_plt.rds")

#Check how many gene are significant
sum(res2$padj < 0.2, na.rm=TRUE)
length(which(res2$pvalue < 0.01))


## MA plot 
DESeq2::plotMA(res2)

# Volcano plot
volcano_plot <- function(res2, threshold = 0.01, log2FC_threshold = 1) {
  significant <- res2$pvalue < threshold & abs(res2$log2FoldChange) > log2FC_threshold
  plot(res2$log2FoldChange, -log10(res2$pvalue), pch=20, cex=0.5, col=ifelse(significant, "red", "black"),
       main="Volcano Plot", xlab="log2 Fold Change", ylab="-log10(pvalue)")
  abline(h = -log10(threshold), col="blue", lty=2)
  abline(v = c(-log2FC_threshold, log2FC_threshold), col="blue", lty=2)
}

# Create the volcano plot
volcano_plot(res2)

colData(dds2)
d <- plotCounts(dds2, gene=which.min(res2$padj), intgroup="age", 
                returnData=TRUE)
ggplot(d, aes(x=age, y=count)) + 
  geom_point(position=position_jitter(w=0.1,h=0)) + 
  scale_y_log10(breaks=c(25,100,400))


vsd <- vst(dds2)

select <- order(rowMeans(counts(dds2,normalized=TRUE)),
                decreasing=TRUE)[1:20]
df <- as.data.frame(colData(dds2)[,c("age")])
pheatmap(assay(vsd)[select,], cluster_rows=FALSE, show_rownames=FALSE,
         cluster_cols=FALSE, annotation_col=df)

list.gene <- get_fc_list(res2)
df.enri <- rna_enrichment(list.gene,  "/gpfs/data/abl/home/cather01/rugglesLab/PACE/enrich_sex")
df.gsea <- gsea_analysis(list.gene,  "/gpfs/data/abl/home/cather01/rugglesLab/PACE/gsea_analysis_sex")


res.df <- res2 %>%
  data.frame() %>%
  rownames_to_column(var="gene") %>% 
  as_tibble()

sig<- res.df %>%
        filter(padj < 0.2 & abs(log2FoldChange) > 0.45)



data <- as.matrix(sig)
heatmap(data)

```


```{r male_vs_female}
#### Male vs Female###
data = read.csv("/gpfs/data/abl/home/cather01/rugglesLab/plt_vs_wb/countmaxtrix_plt.csv", row.names = 1)
metadata = read.csv("/gpfs/data/abl/home/cather01/rugglesLab/PACE/pace_and_thr_metadata.csv", row.names = 1)
metadata <- metadata[rownames(metadata) %in% colnames(data), ]

p <- ggplot(metadata, aes(x=sex)) + 
  geom_histogram( fill="#69b3a2", color="#e9ecef", alpha=0.9)+theme_bw()+ scale_x_continuous(breaks = seq(0, 100, 5))
names(metadata)
metadata$patient <- rownames(metadata)

# Delete outliers 
metadata <- as.data.frame(metadata[rownames(metadata) != "PACE011", ])
data <- data[, colnames(data) != "PACE011"]

out.path <- "/gpfs/data/abl/home/cather01/rugglesLab/"
tabular.stat.table <- metadata %>%
  dplyr::select("age", "sex", "ethnicity", "smoking", "bmi", "diabetes","CAD","hypertension", "hyperlipidemia") %>%
  kbl("html") %>%
  kable_styling(bootstrap_options = "striped", full_width = F) %>%
  kable_minimal() %>%
  save_kable(paste0(out.path, "stat_table.html"))

#Transform column into factor
met <- as.data.frame(metadata$sex)
rownames(met)<-rownames(metadata)
colnames(met) <- c("sex")
met$sex <- as.factor(met$sex)


female_data <- met[met$sex == "Female", ]
male_data <- met[met$sex == "Male", ]
length(female_data)
length(male_data)


#Verification
all(colnames(data) %in% rownames(met))
all(colnames(data) == rownames(met))

#Construct DESEQ2 object
dds <- DESeqDataSetFromMatrix(countData = data, 
                               colData = met,
                               design = ~ sex)
dds


# Filtering 
dds2 <- rna_preprocessing(dds, "/gpfs/data/abl/home/cather01/rugglesLab/PACE/plt_deseq_preprocess")
dds2
# Plot histogram with ggplot2
#Run DESeq
dds2 <- DESeq(dds2)
res2 <- results(dds2)
summary(res2)

max(res2$lfcSE)
res2 <- res2[res2$log2FoldChange < 9.947548,]
# saveRDS(dds2, "/gpfs/data/abl/home/cather01/rugglesLab/plt_vs_wb/dds_plt.rds")

#Check how many gene are significant
sum(res2$padj < 0.1, na.rm=TRUE)
length(which(res2$pvalue < 0.01))


## MA plot 
DESeq2::plotMA(res2)

# Volcano plot
volcano_plot <- function(res2, threshold = 0.05, log2FC_threshold = 1) {
  significant <- res2$pvalue < threshold & abs(res2$log2FoldChange) > log2FC_threshold
  plot(res2$log2FoldChange, -log10(res2$pvalue), pch=20, cex=0.5, col=ifelse(significant, "red", "black"),
       main="Volcano Plot", xlab="log2 Fold Change", ylab="-log10(pvalue)")
  abline(h = -log10(threshold), col="blue", lty=2)
  abline(v = c(-log2FC_threshold, log2FC_threshold), col="blue", lty=2)
}


# Create the volcano plot
volcano_plot(res2)

resultsNames(dds2)
res_tableOE <- lfcShrink(dds2, coef="sex_Male_vs_Female")

EnhancedVolcano(toptable = res_tableOE,
                x = "log2FoldChange",
                y = "padj",
                lab = rownames(res_tableOE),
                xlim = c(-10, +10),
                ylim = c(0, 3),
                pCutoff = 0.2,
                pointSize = 2.0,
                title = "Pseudomonas syringae DC3000 versus mock \n (fold change cutoff = 2, p-value cutoff = 1e-06)"
                # legend=c(
                #   'Not significant',
                #   'Log2 fold-change (but do not pass p-value cutoff)',
                #   'Pass p-value cutoff',
                #   'Pass both p-value & Log2 fold change')
                )




DESeq2::plotMA(res2, ylim=c(-2,5))
DESeq2::plotMA(res_tableOE, ylim=c(-2,5))

res_tableOE_tb <- res_tableOE %>%
  data.frame() %>%
  rownames_to_column(var="gene") %>% 
  as_tibble()

sigOE <- res_tableOE_tb %>%
        filter(padj < 0.2 & abs(log2FoldChange) > 0.45)

list.gene <- get_fc_list(res2)
df.enri <- rna_enrichment(list.gene,  "/gpfs/data/abl/home/cather01/rugglesLab/PACE/enrich")
df.gsea <- gsea_analysis(list.gene,  "/gpfs/data/abl/home/cather01/rugglesLab/PACE/gsea_analysis")

```


