---
title: "harp_pltVSwb_analysis"
output: html_document
date: "2023-11-27"
---

```{r setup, include=FALSE}
packages.list <- c("psych", "clusterProfiler", "ggplot2", "smplot2", "colorRamp2", "dplyr", "grDevices", "gt", "kableExtra","VennDiagram", "readxl", "ggalluvial", "RColorBrewer","viridis","DOSE", "org.Hs.eg.db", "enrichR", "reader")
for (package in packages.list) {
  if (!require(package, character.only = TRUE)) {
    install.packages(package)
  }
}
sapply(packages.list, library, character.only = TRUE)

source("/gpfs/home/cather01/platelet/PLATELET/bulkRNA_analysis.R")
# source("/gpfs/home/cather01/platelet/ToshCode/rnaseq_processing_functions.R")
# source("/gpfs/home/cather01/platelet/ToshCode/mgc_plotting_functions.R")
# source("/gpfs/home/cather01/platelet/ToshCode/geneset_analysis_functions.R")
#install.packages("devtools")
#devtools::install_github('smin95/smplot2')
```


```{r paths}
out.path = "/gpfs/data/abl/home/cather01/rugglesLab/poubelle"

out.path.harp <- "/gpfs/data/abl/home/cather01/rugglesLab/HARP/"
out.path.pace <- "/gpfs/data/abl/home/cather01/rugglesLab/PACE/"

# Define color palette for plots
pal <- c("#CAE1FF", "#20B2AA", "#FFA500", "#DC143C", "#8FBC8F", "#36648B", "#8B8B83", "#FFB6C1", "#DA70D6", "#CD5B45", "#FFD39B", "#BDFCC9", "#698B22", "#FFC125", "#AB82FF", "#5E2612", "#F0E68C", "#4876FF", "#00CED1", "#1B9E77", "#D95F02", "#7570B3", "#E7298A", "#66A61E", "#E6AB02", "#A6761D", "#666666","#E41A1C", "#377EB8", "#4DAF4A", "#984EA3", "#FF7F00")


```

```{r load}
# Read counts data and metadata for HARP cohort
wb.count.harp <- read.table(paste0(out.path.harp, "DATA/wb_quant.featurecounts.counts.unstr.txt"), sep = "\t", header = TRUE, row.names = 1, comment.char = "")
plt.count.harp <- read.table(paste0(out.path.harp,"DATA/plt_quant.featurecounts.counts.unstr.txt"), sep = "\t", header = TRUE, row.names = 1, comment.char = "")
metadata.harp <- read_excel(paste0(out.path.harp,"DATA/Metadata HARP for RNAseq 12_12_2022x.xlsx"))

# Read counts data and metadata for PACE cohort
wb.count.pace <- read.table(paste0(out.path.pace,"DATA/pace_wb_quant.featurecounts.counts.unstr.txt"), sep = "\t", header = TRUE, row.names = 1, comment.char = "")
plt.count.pace <- read.table(paste0(out.path.pace,"DATA/pace_platelet_quant.featurecounts.counts.unstr.txt"), sep = "\t", header = TRUE, row.names = 1, comment.char = "")
metadata.pace <- read.csv(paste0(out.path.pace,"DATA/pace_and_thr_metadata.csv"))

```



```{r preprocessing}

# Function for preprocessing and normalization
preprocess_and_normalize <- function(count.table, tissue, sample_outfilepath, label_extreme_changes) {
  # Create output directory
  sample_outfilepath <- paste0(out.path, sample_outfilepath)
  dir.create(sample_outfilepath, showWarnings = FALSE, recursive = TRUE)  # Filter for Read Count
  list.filter.read <- QC.filter.readcount(count.table, 1000000,sample_outfilepath )
  print("QC reads done")

  # Filter for Gene Count
  minsamplescutoff <- round(ncol(list.filter.read$count.tab.filt)/2)
  list.filter.gene <-  QC.filter.genecount(list.filter.read$count.tab.filt, 8, minsamplescutoff,sample_outfilepath)
  print("QC genes done")
 
  # Create dummy metadata
  dummymeta <- data.frame(tissue = rep(tissue, ncol(list.filter.gene$count.tab.filt)), row.names = colnames(list.filter.gene$count.tab.filt))

    # DESeq normalization
  normalization_out <-DEseq.Normalization(list.filter.gene$count.tab.filt, dummymeta, sample_outfilepath)
  print("DESeq2 done")
 
  # Manually remove extreme outliers
  normcounttab <- normalization_out[[1]][,!colnames(normalization_out[[1]]) %in% rownames(normalization_out$deseqnorm_outliers)]
  return(normcounttab)
}

# Preprocess and normalize for HARP
wb.norm.count.harp <- preprocess_and_normalize(wb.count.harp, "WB", "harp_preprocessing/harp_wb/", FALSE)
plt.norm.count.harp <- preprocess_and_normalize(plt.count.harp, "platelet", "harp_preprocessing/harp_platelet/", FALSE)

# Preprocess and normalize for PACE
wb.norm.count.pace <- preprocess_and_normalize(wb.count.pace, "WB", "pace_preprocessing/pace_wb/", FALSE)
plt.count.pace <- plt.count.pace[,grepl("PACE", colnames(plt.count.pace))] ## GRABBING ONLY PACE SAMPLES
plt.norm.count.pace <- preprocess_and_normalize(plt.count.pace, "platelet", "pace_preprocessing/pace_platelet/", FALSE)


dim(plt.norm.count.pace)
dim(wb.norm.count.pace)
```

```{r sharedFeatures}
############### FOR PACE ###############
colnames(plt.norm.count.pace) <- gsub("_", "", colnames(plt.norm.count.pace))

# Select shared samples and genes, and then select the corresponding tables
# Shared samples are those present in both wb.norm.count and plt.norm.count
# Shared genes are those present in both plt.norm.count and wb.norm.count
shared.samples.pace <- intersect(colnames(wb.norm.count.pace), colnames(plt.norm.count.pace))
shared.genes.pace <- intersect(rownames(plt.norm.count.pace), rownames(wb.norm.count.pace))

# Create matrices containing only shared samples and genes
shared.wb.mtx.pace <- wb.norm.count.pace[shared.genes.pace, shared.samples.pace]
shared.plt.mtx.pace <- plt.norm.count.pace[shared.genes.pace, shared.samples.pace]

################# FOR HARP #################
# Replace underscores in column names with dots
colnames(wb.norm.count.harp) <- gsub("_", ".", colnames(wb.norm.count.harp))
shared.samples.harp <- intersect(colnames(wb.norm.count.harp), colnames(plt.norm.count.harp))
shared.genes.harp <- intersect(rownames(plt.norm.count.harp), rownames(wb.norm.count.harp))

shared.wb.mtx.harp <- wb.norm.count.harp[shared.genes.harp, shared.samples.harp]
shared.plt.mtx.harp <- plt.norm.count.harp[shared.genes.harp, shared.samples.harp]



# Create a directory for HARP analyses (if it doesn't exist)
dir.create(paste0(out.path, "harp_analyses/"), showWarnings = FALSE, recursive = TRUE)
dir.create(paste0(out.path, "pace_analyses/"), showWarnings = FALSE, recursive = TRUE)

# Write shared matrices to CSV files
write.table(shared.wb.mtx.harp, 
            paste0(out.path, "harp_analyses/", "harp_wb_shared_normcounttab.csv"), sep = ",", col.names = NA, row.names = TRUE)
write.table(shared.plt.mtx.harp, 
            paste0(out.path, "harp_analyses/", "harp_platelet_shared_normcounttab.csv"), sep = ",", col.names = NA, row.names = TRUE)

write.table(shared.wb.mtx.pace, 
            paste0(out.path, "pace_analyses/", "pace_wb_shared_normcounttab.csv"), sep = ",", col.names = NA, row.names = TRUE)
write.table(shared.plt.mtx.pace, 
            paste0(out.path, "pace_analyses/", "pace_platelet_shared_normcounttab.csv"), sep = ",", col.names = NA, row.names = TRUE)

dim(shared.plt.mtx.harp)
dim(shared.wb.mtx.pace)

```

```{r sharedAnalysis}
path.harp <- cat.path(out.path, "HARP")
path.pace <- cat.path(out.path, "PACE")
dir.create(path.harp, showWarnings = FALSE, recursive = TRUE)  # Filter for Read Count
dir.create(path.pace, showWarnings = FALSE, recursive = TRUE)  # Filter for Read Count

# Calculate gene metrics for PACE and HARP
gene.out.df.pace <- calculate.gene.metrics(shared.genes.pace, shared.wb.mtx.pace, shared.plt.mtx.pace, path.pace)
gene.out.df.harp <- calculate.gene.metrics(shared.genes.harp, shared.wb.mtx.harp, shared.plt.mtx.harp, path.harp)
```



```{r overall}
overall_expression <- function(df){
  # Ad hoc function to bin genes based on log2(expression) thresholds
  bin_genes <- function(df, thresholds, col) {
    bins <- cut(df[[col]], breaks = c(-Inf, thresholds, Inf), labels = FALSE, include.lowest = TRUE)
    df$ExpressionBin <- bins
    return(df)
  }
  
  plot_bin <- function(df, tissue){
    bin.summary <- df %>%
    group_by(ExpressionBin) %>%
    summarise(mean_Rval = mean(Rval),
              log2_expression = mean(avg_log2)) 
    
    # Plot using ggplot
    p <- (ggplot(bin.summary, aes(x = log2_expression, y = mean_Rval)) 
                    + geom_point() 
                    + labs(x = 'Expression Bin', 
                          y = 'Mean Rval', 
                          title = paste0(tissue, ": Correlations over a range of log2(Expression) thresholds"))
                    + theme_minimal())
    
    p2 <- df %>%
          mutate( bin=cut_width(avg_log2, width=1, boundary=0) ) %>%
          ggplot( aes(x=bin, y=Rval) ) +
              geom_boxplot(fill="#69b3a2") +
              xlab(tissue)+
              ggtitle(paste0(tissue, ": Overall expression")) +
              theme_minimal()
    plots <- list(p, p2)
    names(plots) <- c(paste0(tissue, "point"), paste0(tissue, "box"))
    return(plots)
  }
  # For example, thresholds <- seq(0, 10, by = 1)
  print(paste0("max: ", max(df$plt_avg_log2)))
  thresholds <- seq(min(df$plt_avg_log2)-3, max(df$plt_avg_log2)+3, by = 0.1)
  df.bin.wb <- bin_genes(df, thresholds,"wb_avg_log2")
  df.bin.wb$avg_log2 <- df.bin.wb$wb_avg_log2

  df.bin.plt <- bin_genes(df, thresholds,"plt_avg_log2")
  df.bin.plt$avg_log2 <- df.bin.plt$plt_avg_log2

  plots <- plot_bin(df.bin.wb, "WB")
  plots2 <- plot_bin(df.bin.plt, "PLT")
  plots <- c(plots, plots2)
  return(plots)
    
}

plots.pace <- overall_expression(gene.out.df.pace)
for(plot.name in names(plots.pace)){
  p <- plots.pace[[plot.name]]
  save_plot(p, "pace_analysis/", paste0(plot.name, "_plot_overall_expression.pdf"))
}
plots.harp <- overall_expression(gene.out.df.harp)
for(plot.name in names(plots.harp)){
  p <- plots.harp[[plot.name]]
  save_plot(p, "harp_analysis/", paste0(plot.name, "_plot_overall_expression.pdf"))
}

```

```{r topgenes}
library(adlift)


list.top.genes.pace <- analyze.top.genes(gene.out.df.pace, path.pace)
list.top.genes.harp <- analyze.top.genes(gene.out.df.harp,path.harp)

top.genes.plt.harp<- list.top.genes.harp$plt
top.genes.wb.harp<- list.top.genes.harp$wb
top.genes.plt.pace<- list.top.genes.pace$plt
top.genes.wb.pace <- list.top.genes.pace$wb

write.table(top.genes.plt.harp,paste0(out.path.harp, "DATA/top_genes_plt.csv"), sep = ",", row.names = FALSE)
write.table(top.genes.wb.harp,paste0(out.path.harp, "DATA/top_genes_wb.csv"), sep = ",", row.names = FALSE)
write.table(top.genes.plt.pace,paste0(out.path.pace, "DATA/top_genes_plt.csv"), sep = ",", row.names = FALSE)
write.table(top.genes.wb.pace,paste0(out.path.pace, "DATA/top_genes_wb.csv"), sep = ",", row.names = FALSE)

top.gene.enrichment(list.top.genes.pace,path.pace )
top.gene.enrichment(list.top.genes.harp,path.harp )


```




```{r comparisonHARPPACE}

top.gene.plt.com <- top.genes.plt.harp %>% 
   filter(top.genes.plt.harp$gene %in% top.genes.plt.pace$gene)
top.gene.wb.com <- top.genes.wb.harp %>% 
   filter(top.genes.wb.harp$gene %in% top.genes.wb.pace$gene) 

dim(top.gene.plt.com)
dim(top.gene.wb.com)

com.harp.pace <- merge(gene.out.df.harp, gene.out.df.pace, by = "gene", all = TRUE, suffixes = c(".harp", ".pace"))
com.harp.pace <- na.omit(com.harp.pace)
dim(com.harp.pace)


p <- (ggplot(com.harp.pace, aes(x=wb_avg_log2.harp, y=wb_avg_log2.pace)) 
                      + geom_point(alpha=0.3, size=1) 
                      + theme_bw()  
                      + sm_statCorr(corr_method = 'spearman', color = '#8DD3C7')
                      + ggtitle("HARP: Whole blood expression correlate to PACE: Whole blood expression")
                      + xlab("HARP: Whole blood expression (log2)")
                      + ylab("PACE: Whole blood expression (log2)")
)

p2 <- (ggplot(com.harp.pace, aes(x=plt_avg_log2.harp, y=plt_avg_log2.pace)) 
                      + geom_point(alpha=0.3, size=1) 
                      + theme_bw()  
                      + sm_statCorr(corr_method = 'spearman', color = '#8DD3C7')
                      + ggtitle("HARP: Platelet expression correlate to PACE: Platelet expression")
                      + xlab("HARP: Platelet expression (log2)")
                      + ylab("PACE: Platelet expression (log2)")
)

pdf(paste0(out.path, "scatterplot_harp_pace.pdf"))
print(p)
print(p2)
dev.off()

library(RColorBrewer)
myCol <- brewer.pal(3, "Pastel2")[c(1, 2)]

gene.sets <- list(df1 = gene.out.df.harp$gene, df2 = gene.out.df.pace$gene)
venn.plot <- venn.diagram(gene.sets, 
                          category.names = c("HARP", "PACE"),
                          filename = paste0(out.path, "venn_harp_pace.png"), imagetype="png",
                          output = TRUE,
                          fill = myCol,
                          fontfamily = "sans",
                          cat.fontfamily = "sans"
)


gene.sets.top <- list(df1 = top.genes.plt.harp$gene, df2 = top.genes.plt.pace$gene)
gene.sets.top <- list(df1 = gene.out.df.harp$gene, df2 = top.genes.plt.pace$gene)

venn.plot <- venn.diagram(gene.sets.top, 
                          category.names = c("HARP", "PACE"),
                          filename = paste0(out.path, "venn_harp_paceTOP.png"), imagetype="png",
                          output = TRUE,
                          fill = myCol,
                          fontfamily = "sans",
                          cat.fontfamily = "sans"
)


p <- (ggplot(com.harp.pace, aes(x=Rval.harp, y=Rval.pace)) 
                      + geom_point(alpha=0.3, size=1) 
                      + theme_bw()  
                      + sm_statCorr(corr_method = 'spearman', color = '#8DD3C7')
                      + ggtitle("Correlation of platelet and whole blood correlation for HARP and PACE cohorts")
                      + xlab("HARP: R value of whole blood VS platelet expression")
                      + ylab("PACE: R value of whole blood VS platelet expression")
)

```
```{r}
# Assuming com.harp.pace is your data frame containing the correlation values

# Create a new variable indicating the direction of correlation
com.harp.pace$direction <- ifelse(com.harp.pace$Rval.harp > 0 & com.harp.pace$Rval.pace > 0, "both_up",
                                  ifelse(com.harp.pace$Rval.harp < 0 & com.harp.pace$Rval.pace < 0, "both_down",
                                         ifelse(com.harp.pace$Rval.harp > 0 & com.harp.pace$Rval.pace < 0, "harp_up_pace_down",
                                                ifelse(com.harp.pace$Rval.harp < 0 & com.harp.pace$Rval.pace > 0, "harp_down_pace_up", "uncategorized"))))

# Create a color palette based on the direction
color_palette <- c("both_up" = "#36648B", "both_down" = "#20B2AA", "harp_up_pace_down" = "#FFA500", "harp_down_pace_up" = "#DC143C", "uncategorized" = "gray")

# Create the modified plot
p <- ggplot(com.harp.pace, aes(x = Rval.harp, y = Rval.pace, color = direction)) +
  geom_point(alpha = 0.3, size = 1) +
  scale_color_manual(values = color_palette) +  # Set the color palette
  theme_bw() +
  sm_statCorr(corr_method = 'spearman', color = '#8DD3C7') +
  ggtitle("Correlation of platelet and whole blood correlation for HARP and PACE cohorts") +
  xlab("HARP: R value of whole blood VS platelet expression") +
  ylab("PACE: R value of whole blood VS platelet expression") +
  theme(axis.line = element_line(color = "black", size = 0.5),  # Add axis lines
        axis.text = element_text(size = 10),  # Adjust axis text size
        axis.title = element_text(size = 12, face = "bold"))  # Adjust axis title size

# Add axis at 0,0
p <- p + geom_hline(yintercept = 0, linetype = "dashed", color = "black") +
  geom_vline(xintercept = 0, linetype = "dashed", color = "black")

# Identify the top correlated genes
top_genes <- com.harp.pace %>%
  arrange(desc(abs(Rval.harp * Rval.pace))) %>%
  head(10)

# Add gene names next to their corresponding points
p <- p + geom_text(data = top_genes, aes(x = Rval.harp, y = Rval.pace, label = gene), size = 3, color = "black", hjust = -0.2, vjust = 0)

# Print the plot
print(p)



```

```{r}

gene.out.df.harp[, "log2expressiondiff"] <- log2(gene.out.df.harp[,"plt_avg"]/gene.out.df.harp[,"wb_avg"])
## write out the table

gene.out.df.harp[,"pvalsig"] <- ifelse(gene.out.df.harp[,"Pval"] < 0.05 & gene.out.df.harp[,"Rval"] > 0, "darkred",
                            ifelse(gene.out.df.harp[,"Pval"] < 0.05 & gene.out.df.harp[,"Rval"] < 0, "lightblue", "grey"))


phist <- (ggplot(gene.out.df.harp, aes(x = Rval, fill = pvalsig))
                    + scale_fill_manual(breaks = c("darkred", "grey", "lightblue"), 
                                        values = c("darkred", "grey", "lightblue"))
                    + geom_histogram(color = "black", bins = 100, alpha = 1, position="stack")
                    + labs(title ="Rvals for WB vs Platelet Expression", fill = "significance" )
                    + theme_bw()
)


sum(gene.out.df.harp$pvalsig == "darkred")
sum(gene.out.df.harp$pvalsig == "lightblue")
sum(gene.out.df.harp$pvalsig == "grey")

```

```{r}

plottable <- data.frame(apply(gene.out.df.pace[,c("wb_avg", "plt_avg", "Rval", "Pval")], 2, as.numeric))
plottable[,c("wb_avg", "plt_avg")] <- log2(plottable[,c("wb_avg", "plt_avg")]+1)
rownames(plottable) <- gene.out.df.pace$gene

cortab <- plottable
cortab[,"p.adj"] <- p.adjust(cortab[,"Pval"], method = "fdr")
topcortab <- plottable[plottable[,grepl("Pval", colnames(plottable))] < 0.05,]
topcortab <- topcortab[order(topcortab[,grepl("Rval", colnames(topcortab))], decreasing = TRUE),]

speciesparam <- "Homo sapiens"
m_t2g_cat <- as.data.frame(msigdbr(species = speciesparam, category = c("C5"))[,c("gs_name", "gene_symbol")])
enricheroutUP <- enricher(gene = rownames(topcortab[topcortab[,"Rval"] > 0,]), 
                          qvalueCutoff = 2, pvalueCutoff = 2, minGSSize = 0, maxGSSize = 10000, TERM2GENE = m_t2g_cat)
enricheroutDN <- enricher(gene = rownames(topcortab[topcortab[,"Rval"] < 0,]), 
                          qvalueCutoff = 2, pvalueCutoff = 2, minGSSize = 0, maxGSSize = 10000, TERM2GENE = m_t2g_cat)

class(enricheroutUP)
names(enricheroutUP)
head(enricheroutUP@result, 10)
plotEnrich(enricheroutUP@result, showTerms = 20, numChar = 50, y = "Count", orderBy = "P.value")
barplot(
  enricheroutUP,
  x = "Count",
  color = "p.adjust",
  showCategory = 8,
  font.size = 12,
  title = "",
  label_format = 30,
)

```


```{r}

outfilepathpace <- "/gpfs/data/abl/home/cather01/rugglesLab/testquantileHARP"

## Ok, so a fundamental question here is what genes are captured in platelet sequencing that are NOT in whole blood
# So which genes are seen abundantly in platelets, but NOT abundantly in WB
dir.create(paste0(outfilepathpace, "wb_v_platelet_quantile_analysis/"), showWarnings = FALSE, recursive = TRUE)
quantilegenetable <- gene.out.df.harp
rownames(quantilegenetable) <- quantilegenetable$gene
## Quantile analysis for WB
quantilegenetable <- within(quantilegenetable, wb_quantile<- as.integer(cut(quantilegenetable[,"wb_avg"], 
                                                                            quantile(quantilegenetable[,"wb_avg"], probs=0:20/20, na.rm = TRUE), 
                                                                            include.lowest=TRUE)))
wb_quantiles <- levels(cut(quantilegenetable[,"wb_avg"], 
                           quantile(quantilegenetable[,"wb_avg"], probs=0:20/20, na.rm = TRUE), 
                           include.lowest=TRUE))
## Quantile analysis for platelet
quantilegenetable <- within(quantilegenetable, platelet_quantile <- as.integer(cut(quantilegenetable[,"plt_avg"], 
                                                   quantile(quantilegenetable[,"plt_avg"], probs=0:20/20, na.rm = TRUE), 
                                                   include.lowest=TRUE)))
platelet_quantiles <- levels(cut(quantilegenetable[,"plt_avg"], 
                                quantile(quantilegenetable[,"plt_avg"], probs=0:20/20, na.rm = TRUE), 
                                include.lowest=TRUE))

## Create a table of quantile pairs to plot
quantilecomptable <- as.matrix(table(quantilegenetable[,c("platelet_quantile","wb_quantile")]))
colnames(quantilecomptable) <- paste0("wb_q", colnames(quantilecomptable))
rownames(quantilecomptable) <- paste0("p_q", rownames(quantilecomptable))
quantilecomptable <- quantilecomptable[rev(rownames(quantilecomptable)),]

heatmapcolorparam <- colorRamp2(breaks = c(0, 10, 50, 100), c("white", "#ffcccc", "#ff0000", "#b20000"))
hmout1 <- create_heatmap(counttab = quantilecomptable, scale_data = FALSE, heatmapcolorparam = heatmapcolorparam, addborders = TRUE)
pdf(paste0(outfilepathpace, "wb_v_platelet_quantile_analysis/", "wb_v_platelet_quantile20_hm.pdf"), useDingbats = FALSE, width = 10, height = 10)
draw(hmout1[[1]])
junk <- dev.off()
write.table(quantilecomptable, paste0(outfilepathpace, "wb_v_platelet_quantile_analysis/", "wb_v_platelet_quantile20_table.csv"), sep = ",", col.names = NA, row.names = TRUE)
write.table(quantilegenetable, paste0(outfilepathpace, "wb_v_platelet_quantile_analysis/", "wb_v_platelet_stattable_wquantile.csv"), sep = ",", col.names = TRUE, row.names = FALSE)


## What is the GSEA for the top 25% of both - the shared corners:
genesel <- rownames(quantilegenetable[quantilegenetable[,"platelet_quantile"] >= 16 & quantilegenetable[,"wb_quantile"] >= 16, ])
hypergeo_gsea_out <- hypergeo_genetest(DEseqtable = data.frame(genesel), statcutoffparam = statcutoffparamlist,
                                      genesetparam = c("C5"), speciesparam = "Homo sapiens")
hypergeo_outtable <- data.frame(hypergeo_gsea_out[[1]]) ## fucking beautiful - platelets gene sets right at the top
write.table(hypergeo_outtable, paste0(outfilepathpace, "wb_v_platelet_quantile_analysis/", "wb_v_platelet_hypergeo_topquartile_table.csv"), 
            sep = ",", col.names = TRUE, row.names = FALSE)



## So what are these genes that are in the top 1/2 (??) of platelet genes, but lower 1/2 (??) of WB genes
plateletminquantile <- c(11, 16)
wbmaxquantile <- c(5)
quantilecombos <- expand.grid(wbmaxquantile, plateletminquantile)
colnames(quantilecombos) <- c("wbmaxquantile", "plateletminquantile")
for (quantilecombonum in seq_len(nrow(quantilecombos))) {
  ## Select quantile cut offs for wb and platelet
  plateletminquantile_sel <- quantilecombos[quantilecombonum,"plateletminquantile"]
  wbmaxquantile_sel <- quantilecombos[quantilecombonum,"wbmaxquantile"]
  
  ## Create outpath and select our table and genes
  outpathsel <- paste0(outfilepathpace,  "wb_v_platelet_quantile_analysis/", "pq", plateletminquantile_sel, "_wq", wbmaxquantile_sel, "/")
  dir.create(outpathsel, showWarnings = FALSE, recursive = TRUE)
  tablesel <- quantilegenetable[quantilegenetable[,"platelet_quantile"] >= plateletminquantile_sel &
                                quantilegenetable[,"wb_quantile"] <= wbmaxquantile_sel, ]
  write.table(tablesel, paste0(outpathsel, "wb_v_platelet_statable_sel.csv"), sep = ",", col.names = TRUE, row.names = FALSE)
  genesel <- rownames(tablesel)
  print(length(genesel))
  
  ## Definitely want a gsea for this
  statcutoffparamlist <- "dummy"
  hypergeo_gsea_out = hypergeo_genetest(DEseqtable = data.frame(genesel), statcutoffparam = statcutoffparamlist, 
                                                 genesetparam = c("C5"), speciesparam = "Homo sapiens")
  hypergeo_outtable <- data.frame(hypergeo_gsea_out[[1]]) ## fucking beautiful - platelets gene sets right at the top
  write.table(hypergeo_outtable, paste0(outpathsel, "hypergeo_C5_GOI.csv"), sep = ",", col.names = TRUE, row.names = FALSE)
  
  ## We want a vizualiation of these genes... and maybe with the correlation values - cause if a gene correlates well, then its not so bad, but if it also correlates poorly, than that matters
  # So i guess first - lets just get a scatter of these genes colored by the Rvalue
  pout1 <- scatter_plotter(indata = log2(tablesel[,c("wb_avg", "plt_avg")]+1), colorvar = tablesel[,"Rval",drop=FALSE], plotstats = TRUE,
                           labsparam = list(title = "PACE WB vs Platelet Expression", x = "WB Expression", y = "Platelet Expression"))
  pout1 <- pout1 + scale_color_gradient2(low = "darkblue", mid = "white", high = "darkred", limits = c(-1,1))
  # pout1 <- pout1 + scale_color_gradient(low = "darkblue", high = "darkred", limits = c(-1,1))
  pdf(paste0(outpathsel, "wb_v_platelet_w_rval_scatter.pdf"))
  print(pout1)
  junk <- dev.off()
  
  pout2 <- scatter_plotter(indata = log2(tablesel[tablesel[,"Pval"] < 0.05,c("wb_avg", "plt_avg")]+1), 
                           colorvar = tablesel[tablesel[,"Pval"] < 0.05,"Rval",drop=FALSE], plotstats = TRUE,
                           labsparam = list(title = "PACE WB vs Platelet Expression", x = "WB Expression", y = "Platelet Expression"))
  pout2 <- pout2 + scale_color_gradient2(low = "darkblue", mid = "white", high = "darkred", limits = c(-1,1))
  # pout1 <- pout1 + scale_color_gradient(low = "darkblue", high = "darkred", limits = c(-1,1))
  pdf(paste0(outpathsel, "wb_v_platelet_w_rval_sigonly_scatter.pdf"))
  print(pout2)
  junk <- dev.off()
  
}

```


```{r }
# Function for quantile analysis

perform.quantile.analysis <- function(data){
    quantile.gene.table <- within(data, wb_quantile<- as.integer(cut(quantilegenetable[,"wb_avg"], 
                                                                            quantile(quantilegenetable[,"wb_avg"], probs=0:20/20, na.rm = TRUE), 
                                                                            include.lowest=TRUE)))
wb_quantiles <- levels(cut(quantilegenetable[,"wb_avg"], 
                           quantile(quantilegenetable[,"wb_avg"], probs=0:20/20, na.rm = TRUE), 
                           include.lowest=TRUE))
}

performQuantileAnalysis <- function(data, column, outputDir) {
  # Check if the column contains numeric values
  if (!is.numeric(data[[column]])) {
    stop("The specified column must contain numeric values.")
  }

  # Convert the column to numeric
  data[[column]] <- as.numeric(data[[column]])

  quantiles <- as.integer(cut(data[[column]], quantile(data[[column]], probs = 0:20/20, na.rm = TRUE), include.lowest = TRUE))
  quantileLevels <- levels(cut(data[[column]], quantile(data[[column]], probs = 0:20/20, na.rm = TRUE), include.lowest = TRUE))

  data[[paste0(column, "_quantile")]] <- quantiles
  print(data)
  
  
  quantileComboTable <- table(data$platelet_quantile, data$wb_quantile)
      print(quantileComboTable)

  colnames(quantileComboTable) <- paste0("wb_q", colnames(quantileComboTable))
  
  rownames(quantileComboTable) <- paste0("p_q", rownames(quantileComboTable))
  quantileComboTable <- quantileComboTable[rev(rownames(quantileComboTable)),]

  heatmapColorParam <- colorRamp2(breaks = c(0, 10, 50, 100), c("white", "#ffcccc", "#ff0000", "#b20000"))
  heatmap <- create_heatmap(counttab = quantileComboTable, scale_data = FALSE, heatmapcolorparam = heatmapColorParam, addborders = TRUE)

  pdf(file.path(outputDir, paste0(column, "_quantile_hm.pdf")), useDingbats = FALSE, width = 10, height = 10)
  draw(heatmap[[1]])
  dev.off()

  write.table(quantileComboTable, file.path(outputDir, paste0(column, "_quantile_table.csv")), sep = ",", col.names = NA, row.names = TRUE)
}


# Set the output path
quantile.path <-cat.path(out.path, "quantile")

# Create directory
dir.create(file.path(quantile.path, "wb_v_platelet_quantile_analysis/"), showWarnings = FALSE, recursive = TRUE)

# Perform quantile analysis for WB
performQuantileAnalysis(gene.out.df.harp, "wb.avg", cat.path(quantile.path, "wb_v_platelet_quantile_analysis/"))

# Perform quantile analysis for platelet
performQuantileAnalysis(gene.out.df.harp, "plt_avg", cat.path(quantile.path, "wb_v_platelet_quantile_analysis/"))


```